### PR TITLE
incubator/kube-downscaler - Expand configuration options for service account

### PR DESCRIPTION
**Pull request by patrungel**
#### Is this a new chart
No

#### What this PR does / why we need it:
The present introduces support for imagePullSecrets, annotations, and extra labels for ServiceAccount in use

I moved SA options out of RBAC tree to decouple SA configuration from the one for roles.

I took the liberty of incrementing minor version of the chart as the change is a feature increment.

#### Special notes for your reviewer:
None

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

**Comment by Pluies** [comment]
Hi @patrungel!

This is a breaking change by removing `rbac.serviceAccountName`, so it'd need to be a major version change if we go this route.

Looking at other popular charts (like [cluster-autoscaler](https://github.com/helm/charts/tree/master/stable/cluster-autoscaler), [coredns](https://github.com/helm/charts/tree/master/stable/coredns) or [prometheus-operator](https://github.com/helm/charts/tree/master/stable/prometheus-operator)), it looks like the usual approach to do that would be with:
- `serviceaccount.create` controls whether a new serviceAccount will be created; sometimes just `rbac.create`
- `serviceaccount.name` can be set to an existing serviceAccount name, created outside the chart for full manual control
^ These two appear pretty standard across charts, and I think we should follow this unwritten convention.
- `serviceAccount.annotations`, sometimes `serviceAccountAnnotations` to add annotations on the newly-created serviceAccount

But I haven't seen any other chart that define either labels or pull secrets for automatically-created serviceaccounts. I'd argue that if someone needs that level of customisation, they should probably create the serviceAccount separately and define it with `serviceAccount.name` 👍 

What do you think?

cc @Rowern

**Comment by patrungel** [comment]
Thanks for the review, @Pluies .
- Agree on the breaking character of the change. Is it okay ot up major version, considering the chart is in pre-release (major version 0)? https://semver.org/#how-do-i-know-when-to-release-100
- Confirmed the more adopted layout for serviceAccount to be 
```
create: true
  # The name of the ServiceAccount to use.
  # If not set and create is true, a name is generated using the fullname template
  # name:
```
I'll comply with the above.
- I'd favour `serviceAccount.annotations` over `serviceAccountAnnotations` since we have other `serviceAccount.*` variables.
- Confirmed: `annotations` are rare, `labels` are vertually non-existent (spotted just once really) . I'll drop labels, and leave annotations.
- `imagepullsecrets` is widely supported feature (83 charts seem to have it), but rather as a top-level variable. I can move the variable up the structure and add image pull secrets to the ServiceAccount if created, otherwise to the Deployment. wdyt, @Pluies ?

File: incubator/kube-downscaler/README.md [hunk 1]
```diff
@@ -42,29 +42,32 @@ The command removes all the Kubernetes components associated with the chart and
...
+| `podAnnotations`             | Annotations to be added to downscaler pod                                                               | `{}`                          |
+| `podLabels`                  | Labels to be added to downscaler pod                                                                    | `{}`                          |
+| `resources`                  | Downscaler pod resource requests & limits                                                               | `{}`                          |
+| `securityContext`            | SecurityContext to apply to the downscaler pod                                                          | `{}`                          |
+| `rbac.create`                | If true, create & use RBAC resources                                                                    | `true`                        |
+| `serviceAccount.create`      | If true, create & use a ServiceAccount                                                                  | `""`                          |
```
Comment (Pluies) [inline incubator/kube-downscaler/README.md]: The default here should be `true`

File: incubator/kube-downscaler/templates/serviceaccount.yaml [hunk 2]
```diff
@@ -1,11 +1,16 @@
-{{ if .Values.rbac.create -}}
+{{ if .Values.serviceAccount.create -}}
 apiVersion: v1
 kind: ServiceAccount
 metadata:
-  name: {{ include "kube-downscaler.fullname" . }}
+  name: {{ include "kube-downscaler.serviceAccountName" . }}
+  namespace: {{ .Release.Namespace }}
```
Comment (Pluies) [inline incubator/kube-downscaler/templates/serviceaccount.yaml:6]: Is this line needed? 🤔 I would think it creates resources in the release namespace by default.

**Review by Pluies — CHANGES_REQUESTED** [review]
> Agree on the breaking character of the change. Is it okay ot up major version, considering the chart is in pre-release (major version 0)? https://semver.org/#how-do-i-know-when-to-release-100

I guess that makes sense, especially because we're in the `incubator` part of the charts :)

The imagePullSecrets part looks fine. 👍 

Just a couple of notes on this PR and this is mergeable!

File: incubator/kube-downscaler/templates/serviceaccount.yaml (hunk 2 above)
Comment (patrungel) [reply incubator/kube-downscaler/templates/serviceaccount.yaml:6]: That's an ongoing discussion for about a year or so: https://github.com/helm/helm/issues/5465
While in general helm is supposed to place items into the correct namespace, there are use-cases depending on namespace field to be present. I err on the side of caution here.

**Comment by patrungel** [comment]
> I guess that makes sense, especially because we're in the `incubator` part of the charts :)

Just for the sake of clarity, is it going to be `0.5.0` or `1.0.0`?

File: incubator/kube-downscaler/README.md (hunk 1 above)
Comment (patrungel) [reply incubator/kube-downscaler/README.md]: Fixed that one, thanks!

File: incubator/kube-downscaler/templates/serviceaccount.yaml (hunk 2 above)
Comment (Pluies) [reply incubator/kube-downscaler/templates/serviceaccount.yaml:6]: 🤷 fine; probably should be added to other resources for consistency, but I won't hold the PR based on this!

**Comment by Pluies** [comment]
@patrungel let's keep 0.5.0 👍 

/ok-to-test

/approve

**Comment by scottrigby** [comment]
🚧 This chart is now deprecated.

Further development has moved to [prometheus-community/helm-charts](https://github.com/prometheus-community/helm-charts). The chart has been renamed [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) to more clearly reflect that it installs the `kube-prometheus` project stack, within which Prometheus Operator is only one component.

If applicable, please re-open this PR there. Thanks!

**Comment by Pluies** [comment]
@scottrigby uh, this PR is for the kube-downscaler, as far as I know it has nothing to do with Prometheus – did you actually mean to close it or was it a mistake?

**Comment by scottrigby** [comment]
@Pluies  you're right 👌 This was my mistake somehow from the previous comment, then last night my search picked it up again. Thanks 🙂

**Pull request closed without merging**